app/controllers/schools_controller.rb

```diff
@@ -104,9 +106,127 @@ def update_code_of_conduct
     redirect_to standing_school_path
   end
 
+  # GET /school/discord
+  def discord_configuration
+    authorize current_school
+  end
+
+  # GET /school/discord_server_roles?cause=
+  def discord_server_roles
+    authorize current_school
+
+    if params[:cause].present? && params[:cause] == "DeletedRoles"
+      @sync_service = Discord::SyncRolesService.new(school: current_school)
+    end
+  end
+
+  # PATCH /school/discord_configuration
+  def discord_credentials
+    authorize current_school
+
+    bot_token = params["bot_token"]
+
+    discord_config = current_school.configuration.dig("discord") || {}
+    discord_config["server_id"] = params["server_id"]
+    discord_config["bot_user_id"] = params["bot_user_id"]
+    discord_config["bot_token"] = bot_token.presence ||
+      discord_config.dig("bot_token")
+
+    current_school.update!(
+      configuration:
+        current_school.configuration.merge({ "discord" => discord_config })
+    )
+
+    flash[:success] = I18n.t(
+      "schools.discord_credentials.discord_config_stored"
+    )
+
+    redirect_to discord_configuration_school_path
+  end
+
+  # POST /school/discord_sync_roles
+  def discord_sync_roles
+    authorize current_school
+
+    role_sync_service = Discord::SyncRolesService.new(school: current_school)
+
+    if role_sync_service.sync_ready? &&
+         (!role_sync_service.deleted_roles? || params[:confirmed_sync].present?)
+      role_sync_service.sync
+      flash[:success] = I18n.t(
+        "schools.discord_sync_roles.sync_service_result.success"
+      )
+    elsif role_sync_service.deleted_roles?
+      redirect_to discord_server_roles_school_path(cause: "DeletedRoles")
```

Comment:
This is almost certainly a mistake.

What seems to be happening here is that this route is accepting a request from the user to _sync roles with the Discord server_. While processing that request, it's noticed that we need to inform the user about deleted roles before proceeding.

Instead of rendering a page that shows that information, this redirects the user back to the _Discord roles index page_, and renders a wholly different UI (an issue noted in the related ERB). This redirect necessitates needlessly reloading the information already received from Discord by querying their API again.

Just render the diff page as a response to this request. That way, you don't need to query the Discord server roles information a second time.

If you need to flash notify the user as a part of this response, use `flash.now[TYPE] = ...`.